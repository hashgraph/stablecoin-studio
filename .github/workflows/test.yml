name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test