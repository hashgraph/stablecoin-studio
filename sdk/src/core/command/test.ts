/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
