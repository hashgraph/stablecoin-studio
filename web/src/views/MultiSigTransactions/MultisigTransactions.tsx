import React, { useEffect, useState } from 'react';
import {
	Box,
	Button,
	Modal,
	ModalBody,
	ModalCloseButton,
	ModalContent,
	ModalFooter,
	ModalHeader,
	ModalOverlay,
	Select,
	Table,
	TableContainer,
	Tag,
	Tbody,
	Td,
	Th,
	Thead,
	Tr,
	useDisclosure,
} from '@chakra-ui/react';
// @ts-ignore
import type {
	AccountViewModel,
	MultiSigTransactionsViewModel,
	MultiSigTransactionViewModel,
} from '@hashgraph/stablecoin-npm-sdk';

import { GetTransactionsRequest, SupportedWallets } from '@hashgraph/stablecoin-npm-sdk';
import { ArrowForwardIcon, DeleteIcon } from '@chakra-ui/icons';
import BaseContainer from '../../components/BaseContainer';
import { useTranslation } from 'react-i18next';
import MultiSigTransactionModal from './components/MultiSigTransactionModal';
import SDKService from '../../services/SDKService';
import { useSelector } from 'react-redux';
import { LAST_WALLET_SELECTED, SELECTED_WALLET_ACCOUNT_INFO } from '../../store/slices/walletSlice';

// @ts-ignore
const MultiSigTransactions = () => {
	const { isOpen, onOpen, onClose } = useDisclosure();
	const selectedWallet = useSelector(LAST_WALLET_SELECTED);
	const wallet: AccountViewModel = useSelector(SELECTED_WALLET_ACCOUNT_INFO);
	const publicKey = wallet.publicKey?.key;

	const {
		isOpen: isDeleteModalOpen,
		onOpen: onDeleteModalOpen,
		onClose: onDeleteModalClose,
	} = useDisclosure();
	const [transactionToDelete, setTransactionToDelete] =
		useState<MultiSigTransactionViewModel | null>(null);
	const [transactions, setTransactions] = useState<MultiSigTransactionViewModel[]>([]);
	const [selectedTransaction, setSelectedTransaction] =
		useState<MultiSigTransactionViewModel | null>(null);
	const [filter, setFilter] = useState('');
	const { t } = useTranslation(['multiSig', 'global']);
	const [currentPage, setCurrentPage] = useState(1);
	const [totalPages, setTotalPages] = useState(0);
	const [isSuccessModalOpen, setIsSuccessModalOpen] = useState(false);
	const [successMessage, setSuccessMessage] = useState('');

	useEffect(() => {
		const fetchTransactions = async () => {
			const request = new GetTransactionsRequest({
				limit: 8,
				page: currentPage,
			});

			if (selectedWallet === SupportedWallets.MULTISIG) {
				request.account = wallet.id;
			} else if (publicKey) {
				request.publicKey = { key: publicKey };
			}
			if (request.account || request.publicKey) {
				const resp = await SDKService.getMultiSigTransactions(request);
				setTotalPages(resp.pagination.totalPages);
				setTransactions(resp.transactions);
			}
		};

		fetchTransactions();
	}, [selectedWallet, currentPage]);

	const statusToBgColor = (status: string): string => {
		switch (status) {
			case 'PENDING':
				return 'yellow.200';
			case 'SIGNED':
				return 'green.200';
			case 'EXPIRED':
				return 'orange.200';
			case 'ERROR':
				return 'red.200';
		}
		return 'gray.200';
	};

	const startDateToBgColor = (startDateString: string): string => {
		if (startDateOK(startDateString)) return 'gray.200';
		return 'red.200';
	};

	const startDateOK = (startDateString: string): boolean => {
		const startDate = new Date(startDateString);
		const currentDate = new Date();
		if (startDate > currentDate) return false;
		return true;
	};

	const canSignTransaction = (transaction: MultiSigTransactionViewModel) => {
		return (
			publicKey && transaction.key_list.includes(publicKey) && transaction.status === 'PENDING'
		);
	};

	const canSendTransaction = (transaction: MultiSigTransactionViewModel) => {
		if (selectedWallet === SupportedWallets.METAMASK) return false;
		if (!startDateOK(transaction.start_date)) return false;
		return (
			transaction.signed_keys.length >= transaction.threshold && transaction.status === 'SIGNED'
		);
	};

	const filteredTransactions = transactions.filter((transaction) => {
		if (!filter) return true;
		return transaction.status.toUpperCase() === filter.toUpperCase();
	});

	const handleSignTransaction = async (transactionId: string) => {
		try {
			const response = await SDKService.signMultiSigTransaction(transactionId);
			if (publicKey) {
				setTransactions(
					transactions.map((t) => {
						if (t.id === transactionId) {
							return {
								...t,
								status: 'SIGNED',
								signed_keys: [...t.signed_keys, publicKey],
							};
						}
						return t;
					}),
				);
				console.log('Transaction signed:', response);
			} else {
				console.error('Public key is undefined, cannot sign transaction');
			}
		} catch (error) {
			console.error('Error signing transaction:', error);
		}
	};

	const handleSendTransaction = async (transactionId: string) => {
		try {
			SDKService.submitMultiSigTransaction(transactionId);
			// Remove the transaction from the list as it is sent
			setSuccessMessage(`Operation successfully sent.`);
			setIsSuccessModalOpen(true); // Open the success modal
			setTransactions(transactions.filter((t) => t.id !== transactionId));
		} catch (error) {
			// @ts-ignore
			setSuccessMessage(`Error sending transaction: ${error.message}`);
			setIsSuccessModalOpen(true); // Open the success modal
		}
	};

	const handleConfirmDeleteTransaction = async () => {
		if (transactionToDelete) {
			try {
				await SDKService.removeMultiSigTransaction(transactionToDelete.id);
				setTransactions(transactions.filter((t) => t.id !== transactionToDelete.id));
				onDeleteModalClose();
			} catch (error) {
				console.error('Error deleting transaction:', error);
			}
		}
	};

	const showDeleteConfirmationModal = (transaction: MultiSigTransactionViewModel) => {
		setTransactionToDelete(transaction);
		onDeleteModalOpen();
	};

	const handleDetailsClick = (transaction: MultiSigTransactionViewModel) => {
		setSelectedTransaction(transaction);
		onOpen();
	};

	// @ts-ignore
	return (
		<BaseContainer title={t('Multi-sig transactions')}>
			<Box display='flex' justifyContent='space-between' p={4} bg='white' shadow='sm'>
				<Select
					placeholder='Filter by status'
					width='auto'
					onChange={(e) => setFilter(e.target.value)}
				>
					<option value='pending'>Pending</option>
					<option value='signed'>Signed</option>
				</Select>
			</Box>
			<Box position='relative' mb='4'>
				<TableContainer bg='white' shadow='sm' overflow='hidden'>
					<Table variant='simple'>
						<Thead bg='#ece8ff'>
							<Tr>
								<Th>ID</Th>
								<Th>Description</Th>
								<Th>Account</Th>
								<Th>Threshold</Th>
								<Th>Status</Th>
								<Th>Start Date</Th>
								<Th>Actions</Th>
							</Tr>
						</Thead>
						<Tbody>
							{filteredTransactions.map((transaction) => (
								<Tr key={transaction.id}>
									<Td borderBottom='1px' borderColor='gray.200'>
										{transaction.id}
									</Td>
									<Td
										borderBottom='1px'
										borderColor='gray.200'
										maxWidth='200px'
										whiteSpace='nowrap'
										overflow='hidden'
										textOverflow='ellipsis'
									>
										<span
											style={{ textDecoration: 'underline', cursor: 'pointer' }}
											onClick={() => handleDetailsClick(transaction)}
										>
											{transaction.description}
										</span>
									</Td>
									<Td borderBottom='1px' borderColor='gray.200'>
										{transaction.hedera_account_id}
									</Td>
									<Td borderBottom='1px' borderColor='gray.200'>
										{transaction.threshold}
									</Td>
									<Td borderBottom='1px' borderColor='gray.200'>
										<Tag bg={statusToBgColor(transaction.status)}>{transaction.status}</Tag>
									</Td>
									<Td borderBottom='1px' borderColor='gray.200'>
										<Tag bg={startDateToBgColor(transaction.start_date)}>
											{new Date(transaction.start_date).toDateString() +
												' ' +
												new Date(transaction.start_date).toLocaleTimeString()}
										</Tag>
									</Td>
									<Td borderBottom='1px' borderColor='gray.200'>
										{canSignTransaction(transaction) && (
											<Button
												size={'sm'}
												mr={2}
												style={{ maxWidth: '90px' }}
												rightIcon={<ArrowForwardIcon />}
												onClick={() => handleSignTransaction(transaction.id)}
											>
												Sign
											</Button>
										)}
										{canSendTransaction(transaction) && (
											<Button
												size={'sm'}
												mr={2}
												style={{ maxWidth: '90px' }}
												rightIcon={<ArrowForwardIcon />}
												onClick={() => handleSendTransaction(transaction.id)}
											>
												Send
											</Button>
										)}
										<Button
											size={'sm'}
											mr={2}
											style={{ maxWidth: '90px' }}
											rightIcon={<DeleteIcon />}
											onClick={() => showDeleteConfirmationModal(transaction)}
										>
											Delete
										</Button>
									</Td>
								</Tr>
							))}
						</Tbody>
					</Table>
				</TableContainer>
				{totalPages > 1 && (
					<Box display='flex' justifyContent='center' alignItems='center' p={4}>
						<Button
							size={'sm'}
							style={{ maxWidth: '50px' }}
							onClick={() => setCurrentPage((prevPage) => Math.max(prevPage - 1, 1))}
							disabled={currentPage <= 1}
						>
							Prev
						</Button>
						<Box mx={2}>
							Page {currentPage} of {totalPages}
						</Box>
						<Button
							size={'sm'}
							style={{ maxWidth: '50px' }}
							onClick={() => setCurrentPage((prevPage) => Math.min(prevPage + 1, totalPages))}
							disabled={currentPage >= totalPages}
						>
							Next
						</Button>
					</Box>
				)}
			</Box>
			{selectedTransaction && (
				<MultiSigTransactionModal
					selectedTransaction={selectedTransaction}
					isOpen={isOpen}
					onClose={onClose}
				/>
			)}
			{transactionToDelete && (
				<Modal isOpen={isDeleteModalOpen} onClose={onDeleteModalClose}>
					<ModalOverlay />
					<ModalContent>
						<ModalHeader>Confirm Delete</ModalHeader>
						<ModalCloseButton />
						<ModalBody>Are you sure you want to delete this transaction?</ModalBody>
						<ModalFooter>
							<Button variant='ghost' onClick={onDeleteModalClose}>
								Cancel
							</Button>
							<Button colorScheme='red' mr={3} onClick={handleConfirmDeleteTransaction}>
								Delete
							</Button>
						</ModalFooter>
					</ModalContent>
				</Modal>
			)}
			<Modal isOpen={isSuccessModalOpen} onClose={() => setIsSuccessModalOpen(false)}>
				<ModalOverlay />
				<ModalContent>
					<ModalHeader>Success</ModalHeader>
					<ModalCloseButton />
					<ModalBody>{successMessage}</ModalBody>
					<ModalFooter>
						<Button colorScheme='blue' mr={3} onClick={() => setIsSuccessModalOpen(false)}>
							OK
						</Button>
					</ModalFooter>
				</ModalContent>
			</Modal>
		</BaseContainer>
	);
};

export default MultiSigTransactions;
