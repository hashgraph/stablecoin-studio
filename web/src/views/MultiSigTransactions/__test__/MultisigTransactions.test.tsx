import MultiSigTransactions from '../MultisigTransactions';
import {render} from '../../../test';
import translations from '../../../translations/en/multiSig.json';

describe(`<${MultiSigTransactions.name} />`, () => {

	test('should render correctly', () => {
		const component = render(<MultiSigTransactions />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should have title', () => {
		const component = render(<MultiSigTransactions />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});
});
