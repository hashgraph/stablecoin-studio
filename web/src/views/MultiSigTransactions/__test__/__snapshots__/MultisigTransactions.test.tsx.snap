// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSigTransactions /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-16oka5y"
  >
    <h2
      class="chakra-heading css-1g9h7ui"
      data-testid="base-container-heading"
    >
      Multi-sig transactions
    </h2>
    <div
      class="css-1eyhjq9"
      data-testid="base-container-children"
    >
      <div
        class="css-zttnu1"
      >
        <div
          class="chakra-select__wrapper css-1xrhsyg"
        >
          <select
            class="chakra-select css-7p9xsp"
          >
            <option
              value=""
            >
              Filter by status
            </option>
            <option
              value="signed"
            >
              SIGNED
            </option>
            <option
              value="pending"
            >
              PENDING
            </option>
            <option
              value="expired"
            >
              EXPIRED
            </option>
            <option
              value="error"
            >
              ERROR
            </option>
            <option
              value="executed"
            >
              EXECUTED
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-iohxn1"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="css-zmih7x"
      >
        <div
          class="chakra-table__container css-2wym33"
        >
          <table
            class="chakra-table css-5605sr"
          >
            <thead
              class="css-iqkxbt"
            >
              <tr
                class="css-0"
              >
                <th
                  class="css-1zts0j"
                >
                  ID
                </th>
                <th
                  class="css-1zts0j"
                >
                  Description
                </th>
                <th
                  class="css-1zts0j"
                >
                  Account
                </th>
                <th
                  class="css-1zts0j"
                >
                  Signatures
                </th>
                <th
                  class="css-1zts0j"
                >
                  Threshold
                </th>
                <th
                  class="css-1zts0j"
                >
                  Status
                </th>
                <th
                  class="css-1zts0j"
                >
                  <div
                    class="css-70qvj9"
                  >
                    Start Date
                    <span>
                      <svg
                        class="chakra-icon chakra-icon css-1t4484h"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <g
                          fill="currentColor"
                          stroke="currentColor"
                          stroke-linecap="square"
                          stroke-width="2"
                        >
                          <circle
                            cx="12"
                            cy="12"
                            fill="none"
                            r="11"
                            stroke="currentColor"
                          />
                          <line
                            fill="none"
                            x1="11.959"
                            x2="11.959"
                            y1="11"
                            y2="17"
                          />
                          <circle
                            cx="11.959"
                            cy="7"
                            r="1"
                            stroke="none"
                          />
                        </g>
                      </svg>
                       
                    </span>
                  </div>
                </th>
                <th
                  class="css-1zts0j"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody
              class="css-0"
            />
          </table>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
